### ISSUE TITLE
Error "Argument Y is not a NUMBER" with data only option

### ISSUE THREAD

**Issue by rahul-coatue** [dimitri/pgloader#363]
#### Config without data only
This works just fine
```
LOAD database
  FROM  mysql://user:pass@localhost/test
  INTO  postgresql://user:pass@localhost/test
  WITH  create no tables, create no indexes, include no drop, reset sequences, no foreign keys, downcase identifiers, quote identifiers
  SET   maintenance_work_mem to '512MB', work_mem to '128MB'
;
```

#### Config with data only
Throws the following exception
```
LOAD database
  FROM  mysql://user:pass@localhost/test
  INTO  postgresql://user:pass@localhost/test
  WITH  create no tables, create no indexes, include no drop, reset sequences, no foreign keys, downcase identifiers, quote identifiers, data only
  SET   maintenance_work_mem to '512MB', work_mem to '128MB'
;
```

#### Exception:

```
2016-03-16T13:19:03.024000-07:00 LOG Main logs in '/private/tmp/pgloader/pgloader.log'
2016-03-16T13:19:03.027000-07:00 LOG Data errors in '/private/tmp/pgloader/'
2016-03-16T13:19:03.028000-07:00 LOG Parsing commands from file #P"postgres/pgloader.conf"
An unhandled error condition has been signalled: Argument Y is not a NUMBER: NIL
```

**Comment by rahul-coatue** [comment]
Am I doing something wrong? Please help

**Comment by dimitri** [comment]
I can reproduce it here with data that fails to comply with foreign key constraints. The fix is quite large unfortunately, I will see about that later.

**Comment by pathikrit** [comment]
Can we do a simple fix using psql's single transaction mode? or disable all triggers? Or even just skip the errored rows? Or just print the miscreant rows?

**Comment by pathikrit** [comment]
Why does it happen only with the "data-only" flag and not otherwise?? If there was a FK violation shouldn't without "data-only" fail also?

**Comment by dimitri** [comment]
The problem here is that on foreign key failure, PostgreSQL error message doesn't contain a *context* field in which to read the line number within the batch, which defeats our current batch retry process. This bug only happens in *data-only* situation because when doing the schema and the data, the constraints are put back on top of the data only once the data are all already loaded, so that constraint checking code is done un bulk at the end, not one line at a time during loading. It's much more efficient that way.

I think the right approach here is to compute a topological sort on foreign key dependencies so that we always begin with the reference tables... unless there's a circular dependency on the data.

If you trust your data enough, you can also try to use the `WITH disable triggers` option, so that foreign key constraints are not applied while loading the data.

**Comment by pathikrit** [comment]
> I think the right approach here is to compute a topological sort on foreign key dependencies so that we always begin with the reference tables... unless there's a circular dependency on the data.

Yes, I think my data has circular dependency - I normally do not have a problem with that because I use PostgreSQL's deferred integrity checks. Maybe that is the way to go? By doing things in a single transaction and globally setting the integrity checks to be deferred?

**Comment by pathikrit** [comment]
I just tried with disable triggers and got this:

```
2016-03-23T10:10:24.045000-07:00 LOG Main logs in '/private/tmp/pgloader/pgloader.log'
2016-03-23T10:10:24.055000-07:00 LOG Data errors in '/private/tmp/pgloader/'
2016-03-23T10:10:24.055000-07:00 LOG Parsing commands from file #P"/Volumes/User Data/Users/pbhowmick/Projects/gandalf2/scripts/postgres/pgloader.conf"
An unhandled error condition has been signalled:
   PostgreSQL protocol error: Unexpected message received: Z
```

This is my config:
```
LOAD database
  FROM  mysql://user:pass@localhost/main
  INTO  postgresql://user:pass@localhost/main
  WITH  disable triggers, create no tables, create no indexes, include no drop, reset sequences, no foreign keys, downcase identifiers, quote identifiers, data only
  SET   maintenance_work_mem to '512MB', work_mem to '128MB'
;
```

**Comment by dimitri** [comment]
I can't reproduce here with my local test database. With the same set of options I get the expected unique constraint violations that goes away if I add `truncate` in the WITH option list. Can you retry with the current git version's of pgloader?

**Comment by dimitri** [comment]
The solution has been implemented meanwhile with the non-default *disable triggers* option. Please try again with this option.

**Comment by Chr3is** [comment]
@dimitri what do you mean with "The solution has been implemented meanwhile with the non-default disable triggers option"?

I'm receiving following error:
```
docker run --rm --name pgloader --network=host -v ./pqloader:/data dimitri/pgloader:latest pgloader /data/migration.load
2021-04-08T12:47:11.008000Z LOG pgloader version "3.6.3~devel"
2021-04-08T12:47:11.010000Z LOG Parsing commands from file #P"/data/migration.load"
2021-04-08T12:47:11.061000Z LOG Migrating from #<MYSQL-CONNECTION mysql://root@127.0.0.1:3307/mydb {1006BBDE83}>
2021-04-08T12:47:11.061000Z LOG Migrating into #<PGSQL-CONNECTION pgsql://root@127.0.0.1:5432/mydb {1006BBF263}>

2021-04-08T12:47:11.557000Z ERROR A thread failed with error: PostgreSQL protocol error: Unexpected message received: Z
KABOOM!
FATAL error: PostgreSQL protocol error: Unexpected message received: Z
An unhandled error condition has been signalled:
   PostgreSQL protocol error: Unexpected message received: Z



2021-04-08T12:47:11.577000Z ERROR Date/time: 2021-04-08-12:47!
An unhandled error condition has been signalled:
   PostgreSQL protocol error: Unexpected message received: Z


Backtrace for: #<SB-THREAD:THREAD "lparallel" RUNNING {1007212263}>
0: (TRIVIAL-BACKTRACE:PRINT-BACKTRACE-TO-STREAM #<SB-IMPL::STRING-OUTPUT-STREAM {1005D8A813}>)
1: (TRIVIAL-BACKTRACE:PRINT-BACKTRACE #<CL-POSTGRES::PROTOCOL-ERROR {1005D8A663}> :OUTPUT NIL :IF-EXISTS :APPEND :VERBOSE NIL)
2: ((LAMBDA (CONDITION) :IN PGLOADER.LOAD:COPY-FROM) #<CL-POSTGRES::PROTOCOL-ERROR {1005D8A663}>)
3: (LPARALLEL.KERNEL::CONDITION-HANDLER #<CL-POSTGRES::PROTOCOL-ERROR {1005D8A663}>)
4: (SB-KERNEL::%SIGNAL #<CL-POSTGRES::PROTOCOL-ERROR {1005D8A663}>)
5: (ERROR CL-POSTGRES::PROTOCOL-ERROR :MESSAGE "Unexpected message received: Z")
6: (CL-POSTGRES::SEND-QUERY #<SB-SYS:FD-STREAM for "socket 127.0.0.1:55210, peer: 127.0.0.1:5432" {1005BA1EA3}> #<unavailable argument> #<unavailable argument>)
7: ((LABELS #:G4 :IN CL-POSTGRES:EXEC-QUERY))
8: (PGLOADER.PGSQL:PGSQL-EXECUTE "ALTER TABLE natural_representatives ENABLE TRIGGER ALL;" :LOG-LEVEL :SQL :CLIENT-MIN-MESSAGES NIL :ON-ERROR-STOP T)
9: ((FLET "CLEANUP-FUN-13" :IN PGLOADER.PGCOPY:COPY-ROWS-FROM-QUEUE)) [cleanup]
10: (PGLOADER.PGCOPY:COPY-ROWS-FROM-QUEUE #<PGLOADER.SOURCE.MYSQL:COPY-MYSQL {1008AA5143}> #S(LPARALLEL.VECTOR-QUEUE:VECTOR-QUEUE :IMPL #S(LPARALLEL.VECTOR-QUEUE::RAW-QUEUE :DATA #(#(...) :START 0 :COUNT 343) :LOCK #<SB-THREAD:MUTEX "Anonymous lock" (free)> :NOTIFY-PUSH NIL :NOTIFY-POP NIL) :DISABLE-TRIGGERS T :ON-ERROR-STOP T :COLUMNS NIL)
11: ((FLET "BODY-FN0" :IN LPARALLEL.KERNEL::MAKE-CHANNELED-TASK))
12: ((LAMBDA NIL :IN LPARALLEL.KERNEL::MAKE-CHANNELED-TASK))
13: (LPARALLEL.KERNEL::EXEC-TASK/WORKER #<unavailable argument> #S(LPARALLEL.KERNEL::WORKER :HANDSHAKE/FROM-WORKER #S(LPARALLEL.CONS-QUEUE:CONS-QUEUE :IMPL #S(LPARALLEL.RAW-QUEUE:RAW-QUEUE :HEAD NIL :TAIL NIL) :LOCK #<SB-THREAD:MUTEX #1="Anonymous lock" (free)> :CVAR NIL) :HANDSHAKE/TO-WORKER #S(LPARALLEL.CONS-QUEUE:CONS-QUEUE :IMPL #S(LPARALLEL.RAW-QUEUE:RAW-QUEUE :HEAD NIL :TAIL NIL) :LOCK #<SB-THREAD:MUTEX #1# (free)> :CVAR #<SB-THREAD:WAITQUEUE Anonymous condition variable {1008F04533}>) :EXIT-NOTIFICATION #S(LPARALLEL.CONS-QUEUE:CONS-QUEUE :IMPL #S(LPARALLEL.RAW-QUEUE:RAW-QUEUE :HEAD NIL :TAIL NIL) :LOCK #<SB-THREAD:MUTEX #1# (free)> :CVAR NIL) :THREAD #<SB-THREAD:THREAD "lparallel" RUNNING {1007212263}> :RUNNING-CATEGORY :DEFAULT :INDEX 6 :TASKS #S(LPARALLEL.SPIN-QUEUE:SPIN-QUEUE :HEAD (LPARALLEL.SPIN-QUEUE::DUMMY (#<CLOSURE #2=(LAMBDA NIL :IN LPARALLEL.KERNEL::MAKE-CHANNELED-TASK) {10089FB5FB}> . :DEFAULT) (#<CLOSURE #2# {10089FB62B}> . :DEFAULT) (#<CLOSURE #2# {10089FB65B}> . :DEFAULT) (#<CLOSURE #2# {10089FB68B}> . :DEFAULT) (#<CLOSURE #2# {10089FB6BB}> . :DEFAULT) (#<CLOSURE #2# {10089FB6EB}> . :DEFAULT) . #3=((#<CLOSURE #2# {10089FB71B}> . :DEFAULT))) :TAIL #3#)))
14: (LPARALLEL.KERNEL::WORKER-LOOP #<LPARALLEL.KERNEL:KERNEL :NAME "lparallel" :WORKER-COUNT 8 :USE-CALLER NIL :ALIVE T :SPIN-COUNT 2000 {100720FDF3}> #S(LPARALLEL.KERNEL::WORKER :HANDSHAKE/FROM-WORKER #S(LPARALLEL.CONS-QUEUE:CONS-QUEUE :IMPL #S(LPARALLEL.RAW-QUEUE:RAW-QUEUE :HEAD NIL :TAIL NIL) :LOCK #<SB-THREAD:MUTEX #1="Anonymous lock" (free)> :CVAR NIL) :HANDSHAKE/TO-WORKER #S(LPARALLEL.CONS-QUEUE:CONS-QUEUE :IMPL #S(LPARALLEL.RAW-QUEUE:RAW-QUEUE :HEAD NIL :TAIL NIL) :LOCK #<SB-THREAD:MUTEX #1# (free)> :CVAR #<SB-THREAD:WAITQUEUE Anonymous condition variable {1008F04533}>) :EXIT-NOTIFICATION #S(LPARALLEL.CONS-QUEUE:CONS-QUEUE :IMPL #S(LPARALLEL.RAW-QUEUE:RAW-QUEUE :HEAD NIL :TAIL NIL) :LOCK #<SB-THREAD:MUTEX #1# (free)> :CVAR NIL) :THREAD #<SB-THREAD:THREAD "lparallel" RUNNING {1007212263}> :RUNNING-CATEGORY :DEFAULT :INDEX 6 :TASKS #S(LPARALLEL.SPIN-QUEUE:SPIN-QUEUE :HEAD (LPARALLEL.SPIN-QUEUE::DUMMY (#<CLOSURE #2=(LAMBDA NIL :IN LPARALLEL.KERNEL::MAKE-CHANNELED-TASK) {10089FB5FB}> . :DEFAULT) (#<CLOSURE #2# {10089FB62B}> . :DEFAULT) (#<CLOSURE #2# {10089FB65B}> . :DEFAULT) (#<CLOSURE #2# {10089FB68B}> . :DEFAULT) (#<CLOSURE #2# {10089FB6BB}> . :DEFAULT) (#<CLOSURE #2# {10089FB6EB}> . :DEFAULT) . #3=((#<CLOSURE #2# {10089FB71B}> . :DEFAULT))) :TAIL #3#)))
15: (LPARALLEL.KERNEL::%CALL-WITH-TASK-HANDLER #<unavailable argument>)
16: ((LAMBDA NIL :IN LPARALLEL.KERNEL::CALL-WITH-WORKER-CONTEXT))
17: (LPARALLEL.KERNEL::CALL-WITH-WORKER-CONTEXT #<CLOSURE (LAMBDA NIL :IN LPARALLEL.KERNEL::ENTER-WORKER-LOOP) {10062C62BB}> #<FUNCTION FUNCALL> #<LPARALLEL.KERNEL:KERNEL :NAME "lparallel" :WORKER-COUNT 8 :USE-CALLER NIL :ALIVE T :SPIN-COUNT 2000 {100720FDF3}> #S(LPARALLEL.KERNEL::WORKER :HANDSHAKE/FROM-WORKER #S(LPARALLEL.CONS-QUEUE:CONS-QUEUE :IMPL #S(LPARALLEL.RAW-QUEUE:RAW-QUEUE :HEAD NIL :TAIL NIL) :LOCK #<SB-THREAD:MUTEX #1="Anonymous lock" (free)> :CVAR NIL) :HANDSHAKE/TO-WORKER #S(LPARALLEL.CONS-QUEUE:CONS-QUEUE :IMPL #S(LPARALLEL.RAW-QUEUE:RAW-QUEUE :HEAD NIL :TAIL NIL) :LOCK #<SB-THREAD:MUTEX #1# (free)> :CVAR #<SB-THREAD:WAITQUEUE Anonymous condition variable {1008F04533}>) :EXIT-NOTIFICATION #S(LPARALLEL.CONS-QUEUE:CONS-QUEUE :IMPL #S(LPARALLEL.RAW-QUEUE:RAW-QUEUE :HEAD NIL :TAIL NIL) :LOCK #<SB-THREAD:MUTEX #1# (free)> :CVAR NIL) :THREAD #<SB-THREAD:THREAD "lparallel" RUNNING {1007212263}> :RUNNING-CATEGORY :DEFAULT :INDEX 6 :TASKS #S(LPARALLEL.SPIN-QUEUE:SPIN-QUEUE :HEAD (LPARALLEL.SPIN-QUEUE::DUMMY (#<CLOSURE #2=(LAMBDA NIL :IN LPARALLEL.KERNEL::MAKE-CHANNELED-TASK) {10089FB5FB}> . :DEFAULT) (#<CLOSURE #2# {10089FB62B}> . :DEFAULT) (#<CLOSURE #2# {10089FB65B}> . :DEFAULT) (#<CLOSURE #2# {10089FB68B}> . :DEFAULT) (#<CLOSURE #2# {10089FB6BB}> . :DEFAULT) (#<CLOSURE #2# {10089FB6EB}> . :DEFAULT) . #3=((#<CLOSURE #2# {10089FB71B}> . :DEFAULT))) :TAIL #3#)))
18: ((LAMBDA NIL :IN LPARALLEL.KERNEL::MAKE-WORKER-THREAD))
19: ((LAMBDA NIL :IN BORDEAUX-THREADS::BINDING-DEFAULT-SPECIALS))
20: ((FLET SB-UNIX::BODY :IN SB-THREAD::INITIAL-THREAD-FUNCTION-TRAMPOLINE))
21: ((FLET "WITHOUT-INTERRUPTS-BODY-4" :IN SB-THREAD::INITIAL-THREAD-FUNCTION-TRAMPOLINE))
22: ((FLET SB-THREAD::WITH-MUTEX-THUNK :IN SB-THREAD::INITIAL-THREAD-FUNCTION-TRAMPOLINE))
23: ((FLET "WITHOUT-INTERRUPTS-BODY-1" :IN SB-THREAD::CALL-WITH-MUTEX))
24: (SB-THREAD::CALL-WITH-MUTEX #<CLOSURE (FLET SB-THREAD::WITH-MUTEX-THUNK :IN SB-THREAD::INITIAL-THREAD-FUNCTION-TRAMPOLINE) {7F4D81F4ED6B}> #<SB-THREAD:MUTEX "thread result lock" owner: #<SB-THREAD:THREAD "lparallel" RUNNING {1007212263}>> NIL T NIL)
25: (SB-THREAD::INITIAL-THREAD-FUNCTION-TRAMPOLINE #<SB-THREAD:THREAD "lparallel" RUNNING {1007212263}> NIL #<CLOSURE (LAMBDA NIL :IN BORDEAUX-THREADS::BINDING-DEFAULT-SPECIALS) {100721220B}> NIL)
26: ("foreign function: call_into_lisp")
27: ("foreign function: new_thread_trampoline")


What I am doing here?

PostgreSQL protocol error: Unexpected message received: Z
```

with following config:
```
LOAD DATABASE
     FROM mysql://root:root@127.0.0.1:3307/mydb
     INTO pgsql://root:root@127.0.0.1:5432/mydb
  
  WITH 
      data only,
      disable triggers,
      include no drop,
      create no tables,
      create no indexes,
      no foreign keys,
      reset sequences,
      workers = 8,
      concurrency = 1,
      multiple readers per thread,
      rows per range = 5000,
      batch size = 10MB,
      prefetch rows = 5000

  EXCLUDING TABLE NAMES MATCHING '..some_tables'

  SET PostgreSQL PARAMETERS
      maintenance_work_mem to '128MB',
      work_mem to '12MB'

  SET MySQL PARAMETERS
      net_read_timeout  = '120',
      net_write_timeout = '120'

  CAST 
       type int when (= precision 11) to integer,
       type datetime to timestamp using zero-dates-to-null,
       type timestamp to timestamp using zero-dates-to-null,
       '... some special column casts';
```
